### ISSUE TITLE
Feature request: await listeners before responding

### ISSUE THREAD

**Issue by bkeepers** [gr2m/octokit-webhooks.js-legacy#7]
The current implementation uses `EventEmitter` and does a fire and forget with listeners. It would be nice to have an option to wait for all event listeners to successfully resolve before returning a 200 response. GitHub currently doesn't change behavior on webhooks that return an error, but it at least will make the error known in the webhook UI.

Here is an example of Slacks event api client doing this:
https://github.com/slackapi/node-slack-events-api/blob/479b48cd9c71e2b5da5c42f9281d024d2b0f6ff0/test/integration/test-adapter-options.js#L7-L47

We could use Promises for the event handlers to accomplish the same thing with something like [promise-events](https://www.npmjs.com/package/promise-events).

cc @wilhelmklopp

**Comment by gr2m** [comment]
@bkeepers just so that I understand this correctly, you would like to to wait the server to wait before it responds to the webhook request coming from GitHub until my asynchronous event handler are finished, so that I can return error codes from my custom handlers?

It sounds like what you want are hooks instead of events? 

We could do something like this

```js
const webhooks = require('@octokit/webhooks')

// called with installation webhook payload
// see https://developer.github.com/v3/activity/events/types/#installationevent
webhooks.on('installation', {action, installation, sender} => {
  console.log(`${sender.login} ${action} installation for ${installation.account.login`)
  // octocat deleted installation for octocat
})

// called with installation event payload, same as above
webhooks.hook('installation', async {action, installation, sender} => {
  if (action === 'deleted') {
    return myApp.installations.remove(installation)
  }
  
  const numAvailableInstallations = await myApp.installations.availableFor(sender.id)
  if (numAvailableInstallations === 0) {
    throw Error('Out of installations, please upgrade your plan')
  }

  return myApp.installations.add(installation)
})

http.createServer(function (req, res) {
  webhooks.handler({
    path: '/webhook',
    secret: 'myhashsecret'
  }, req, res))
}).listen(7777)
```

And our implementation would look something like this

```js
try {
  // 1. emits event based on payload in next tick
  // 2. runs all hooks in parallel and awaits their results
  await handle(installationWebhookPayload)
  res.writeHead(200, { 'content-type': 'application/json' })
  res.end('{"ok":true}')
} catch (error) {
  res.writeHead(error.statusCode, { 'content-type': 'application/json' })
  res.end(error.data)
}
```

**Comment by gr2m** [comment]
this will be done via https://github.com/gr2m/octokit-webhooks/pull/3.

I’m closing this issue as I plan to replace the repo
